applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_small_displacement_element.cpp

```diff
@@ -0,0 +1,201 @@
+#include "custom_response_functions/response_utilities/stress_response_definitions.h"
+#include "includes/checks.h"
+
+
+namespace Kratos
+{
+
+AdjointFiniteDifferenceSmallDisplacementElement::AdjointFiniteDifferenceSmallDisplacementElement(Element::Pointer pPrimalElement)
+    : AdjointFiniteDifferencingBaseElement(pPrimalElement)
+{
+}
+
+AdjointFiniteDifferenceSmallDisplacementElement::~AdjointFiniteDifferenceSmallDisplacementElement()
+{
+}
+
+void AdjointFiniteDifferenceSmallDisplacementElement::CalculateStressDisplacementDerivative(const Variable<Vector>& rStressVariable,
+                                                                                            Matrix& rOutput, const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    // First check applicability of this function
+    KRATOS_ERROR_IF(rStressVariable != STRESS_ON_GP)
+        << "AdjointFiniteDifferenceSmallDisplacementElement::CalculateStressDisplacementDerivative: Invalid stress variable! Stress variable not supported for this element!" << std::endl;
+
+    TracedStressType traced_stress_type = static_cast<TracedStressType>(this->GetValue(TRACED_STRESS_TYPE));
+    KRATOS_ERROR_IF(traced_stress_type != TracedStressType::VON_MISES_STRESS)
+        << "AdjointFiniteDifferenceSmallDisplacementElement::CalculateStressDisplacementDerivative: Invalid stress type! Stress type not supported for this element!" << std::endl;
+
+    KRATOS_ERROR_IF(rCurrentProcessInfo.Has(NL_ITERATION_NUMBER))
+        << "This stress displacement derivative computation is only usable for linear cases!" << std::endl;
+
+    // Some working variables
+    const SizeType num_nodes = mpPrimalElement->GetGeometry().PointsNumber();
+    const SizeType dimension = mpPrimalElement->GetGeometry().WorkingSpaceDimension();
+    const SizeType num_dofs_per_node = (mHasRotationDofs) ?  2 * dimension : dimension;
+    const SizeType num_dofs = num_nodes * num_dofs_per_node;
+
+    // Build vector of variables containing the DOF-variables of the primal problem
+    std::vector<VariableComponent<VectorComponentAdaptor<array_1d<double, 3>>>> primal_solution_variable_list;
+    primal_solution_variable_list.reserve(num_dofs_per_node);
+    primal_solution_variable_list.push_back(DISPLACEMENT_X);
+    primal_solution_variable_list.push_back(DISPLACEMENT_Y);
+    primal_solution_variable_list.push_back(DISPLACEMENT_Z);
+
+    std::vector<Matrix> stress_tensor;
+    mpPrimalElement->CalculateOnIntegrationPoints(PK2_STRESS_TENSOR, stress_tensor, rCurrentProcessInfo);
+
+    const unsigned int number_integration_points = stress_tensor.size();
+
+    // Computation of prefactors
+    std::vector<double> sensitivity_prefactors(number_integration_points);
+
+    for(IndexType k = 0; k < number_integration_points; ++k)
+    {
+        Matrix & PK2_k = stress_tensor[k];
+        double radicant = 0.0;
+        radicant += PK2_k(0,0)*PK2_k(0,0) + PK2_k(1,1)*PK2_k(1,1) + PK2_k(2,2)*PK2_k(2,2);
```

Comment:
This will not work for 2D like this. But 2D also requires a bit more thinking and testing. I would want to leave this to a future PR. Instead, a corresponding error is dropped now.